src/mutable_list.jl

```diff
@@ -2,71 +2,95 @@ mutable struct ListNode{T}
     data::T
     prev::ListNode{T}
     next::ListNode{T}
-    function ListNode{T}() where T
+    function ListNode{T}() where {T}
         node = new{T}()
-        node.next = node
-        node.prev = node
+        node.next = node.prev = node
         return node
     end
-    function ListNode{T}(data) where T
-        node = new{T}(data)
+    function ListNode{T}(prev, data, next) where {T}
+        node = new{T}(data, prev, next)
+        prev.next = next.prev = node
         return node
     end
 end
 
 mutable struct MutableLinkedList{T}
     len::Int
     node::ListNode{T}
-    function MutableLinkedList{T}() where T
-        l = new{T}()
-        l.len = 0
-        l.node = ListNode{T}()
-        l.node.next = l.node
-        l.node.prev = l.node
-        return l
+    function MutableLinkedList{T}() where {T}
+        return new{T}(0, ListNode{T}())
     end
 end
 
 MutableLinkedList() = MutableLinkedList{Any}()
-
-function MutableLinkedList{T}(elts...) where T
-    l = MutableLinkedList{T}()
-    for elt in elts
-        push!(l, elt)
-    end
-    return l
-end
+MutableLinkedList(elts...) = MutableLinkedList{eltype(elts)}(elts...)
+MutableLinkedList{T}(elts...) where {T} = append!(MutableLinkedList{T}(), elts)
 
 Base.iterate(l::MutableLinkedList) = l.len == 0 ? nothing : (l.node.next.data, l.node.next.next)
 Base.iterate(l::MutableLinkedList, n::ListNode) = n === l.node ? nothing : (n.data, n.next)
 
 Base.isempty(l::MutableLinkedList) = l.len == 0
 Base.length(l::MutableLinkedList) = l.len
-Base.collect(l::MutableLinkedList{T}) where T = T[x for x in l]
-Base.eltype(::Type{<:MutableLinkedList{T}}) where T = T
+Base.collect(l::MutableLinkedList{T}) where {T} = T[x for x in l]
```

Comment:
Do we need this?
I would have thought it would fall back to this if we didn't define it.